src/VisualStudio/Core/Def/Implementation/Experimentation/KeybindingResetDetector.cs

```diff
@@ -0,0 +1,378 @@
+            return ((OLECMDF)cmds[0].cmdf).HasFlag(OLECMDF.OLECMDF_ENABLED);
+        }
+
+        private void RestoreVsKeybindings()
+        {
+            AssertIsForeground();
+
+            if (_uiShell == null)
+            {
+                _uiShell = _serviceProvider.GetService<IVsUIShell, SVsUIShell>();
+            }
+
+            ErrorHandler.ThrowOnFailure(_uiShell.PostExecCommand(
+                    VSConstants.GUID_VSStandardCommandSet97,
+                    (uint)VSConstants.VSStd97CmdID.CustomizeKeyboard,
+                    (uint)OLECMDEXECOPT.OLECMDEXECOPT_DODEFAULT,
+                    null));
+
+            KeybindingsResetLogger.Log("KeybindingsReset");
+
+            _workspace.Options = _workspace.Options.WithChangedOption(KeybindingResetOptions.NeedsReset, false);
+        }
+
+        private void OpenExtensionsHyperlink()
+        {
+            ThisCanBeCalledOnAnyThread();
+            if (!BrowserHelper.TryGetUri(KeybindingsFwLink, out Uri fwLink))
+            {
+                // We're providing a constant, known-good link. This should be impossible.
+                throw ExceptionUtilities.Unreachable;
+            }
+
+            BrowserHelper.StartBrowser(fwLink);
+
+            KeybindingsResetLogger.Log("ExtensionsLink");
+            _workspace.Options = _workspace.Options.WithChangedOption(KeybindingResetOptions.NeedsReset, false);
+        }
+
+        private void NeverShowAgain()
+        {
+            AssertIsForeground();
+
+            _workspace.Options = _workspace.Options.WithChangedOption(KeybindingResetOptions.NeverShowAgain, true)
+                                                   .WithChangedOption(KeybindingResetOptions.NeedsReset, false);
+            KeybindingsResetLogger.Log("NeverShowAgain");
+
+            // The only external references to this object are as callbacks, which are removed by the Shutdown method.
+            Shutdown();
+        }
+
+        private void InfoBarClose()
+        {
+            AssertIsForeground();
+            _infoBarOpen = false;
+        }
+
+        public int QueryStatus(ref Guid pguidCmdGroup, uint cCmds, OLECMD[] prgCmds, IntPtr pCmdText)
+        {
+            // Technically can be called on any thread, though VS will only ever call it on the UI thread.
+            ThisCanBeCalledOnAnyThread();
+            // We don't care about query status, only when the command is actually executed
+            return (int)OLE.Interop.Constants.OLECMDERR_E_NOTSUPPORTED;
+        }
+
+        public int Exec(ref Guid pguidCmdGroup, uint nCmdID, uint nCmdexecopt, IntPtr pvaIn, IntPtr pvaOut)
+        {
+            // Technically can be called on any thread, though VS will only ever call it on the UI thread.
+            ThisCanBeCalledOnAnyThread();
+            if (pguidCmdGroup == ReSharperCommandGroup && nCmdID >= ResumeId && nCmdID <= ToggleSuspendId)
```

Comment:
The >= / <= seems odd here. My immediate reaction was "oh goodness, what's in the middle?" and had to go look up the constants.

Reply:
I did this to minimize the number ops we're doing. Yes, it's just 3 comparisons vs 2, but this is a priority command handler and I wanted to get onto the background thread ASAP.